### DISCUSSION TITLE
On a Pico W, EInkPIO may not be usable in a loop

### DISCUSSION THREAD

**Discussion by Cheddoleum** [phoreglad/pico-epaper#10] [Q&A] [answered]
There appears to be a known issue with the Pico W where some PIO instruction memory is not released; apparently for reasons relating to the wifi system.  Calling the StateMachine() constructor repeatedly will result in an ENOMEM after a certain number of iterations.

These issues address a manifestation of the problem: https://github.com/micropython/micropython/issues/9003 and 
https://github.com/micropython/micropython/pull/10652 , but mainly just fix the issue as it pertains to soft resets, e.g. in Thonny, and don't address the issue of reinitializing the state machine within a program. 

I'm using your library in a remote temperature/humidity display; it checks the sensor every few minutes and if it needs to, redraws the screen. After a a given redraw it calls .sleep().  The EInkPIO object is constructed locally to this function and so is GC'd afterward.

To reinitialize the screen after a sleep it appears to be necessary to instantiate a EInkPIO object again, which constructs its ._sm StateMachine member in _pio_setup().  After a fairly small number of refreshes, the program dies with an OSError 12 on that line.

For my use case I can get around this by simply using the Eink class instead. (I'm using generated images so the speed/memory difference when rotated isn't a problem; I can just use rotated images instead.)

But without a way to restart the display after a .sleep() without instantiating a leaky-on-the-PicoW StateMachine, there doesn't seem to be a way to use EInkPIO in a loop on the W.

**Comment by phoreglad** [comment]
Can you instantiate the EinkPIO object only once, outside of any functions and use it as a 'global' object (or pass it as an argument to the function)? That would fix the problem. Other possibility is to manually release the PIO program memory at the end of the function to make sure there's space for the next run.

The issues you've linked to don't really apply here as you're not doing a soft reset between the function calls. Without the soft reset even if the StateMachine object gets deleted and collected, the program that it started still runs and takes instruction space.

**Comment by Cheddoleum** [comment]
I've tried your first suggestion, but after a sleep the EInkPIO object doesn't seem to be usable for drawing again; it just hangs when you try.  Waveshare's docs says that to start using the display again after a sleep (assuming you're both referring to the same "sleep" mode) it must be reinitialized.  Your lib doesn't have a public reinitialization function other than the constructor.

(I may have spoken too soon anyway -- I'm now getting a MemoryError a few iterations in [sometimes!] while using EInk, so the OSError 12 (ENOMEM) on EInkPIO instantiation failure may simply have been a symptom of running out of memory some other way) Even more strangely it's difficult to reproduce. Trying to chase it down now but the inconsistency is bothering me: this is not a complicated program **at all**, and one really doesn't expect memory leaks in Python in the ordinary case anyway.  I'll update when I know more)

**Reply by phoreglad** [reply, marked as answer]
Yes, you're right about reinitialisation. I've updated the library to include a public reinit() method. After putting the screen to sleep make sure to call reinit() before next call to show().

As for the error when using EInk class it might be caused by memory fragmentation. The screen requires relatively large framebuffers, so recreating the EInk object several times and running other code in between might not leave enough contiguous memory for the buffers.

**Reply by Cheddoleum** [reply]
Thank you! I'll give it a try soon.  Just traced an instance of the crash to of EInk send where it creates bytes and map objects to handle 90/270 rotated displays. (I was right in the middle implementing the ability to request a rotated image from the provider so I could stop doing that anyway.)

I'll also look into mitigation tactics for memory fragmentation as you suggest. (If all else fails I'll just soft-reset and restart programmatically in the memory exception handler -- the result shouldn't look out of the ordinary to the user. Gruesome hack though.)

**Reply by phoreglad** [reply]
If you initialise the EInkPIO at the very beginning of your program (to make sure there's plenty of free memory available) and reuse it with the help of a new reinit() method, memory fragmentation shouldn't be a problem and your program should work just fine.

Unfortunately EInk class requires quite a lot of memory to handle the rotations reasonably quickly making it more susceptible to running out of memory and fragmentation problems. The PIO variant is much better in this regard and I strongly recommend using it.

**Reply by Cheddoleum** [reply]
After implementing that, it's been running stably for several hours now, a very solid improvement. Thanks very much again.

A couple of questions when/if you have a moment?

<strike>1. Along the same lines as the above: avoiding large dynamic memory allocations in an eternal loop program on a platform with no real memory management, I'd like to be able to reuse FrameBuffers. But the only obvious way to populate a FrameBuffer with binary data (such as generated by your image-to-py conversion utility) is to do it at construct time; so I'm creating a new one every time I write a new image.*  Is there a way to declare a single FrameBuffer instance globally and reuse it, writing new binary data into it?</strike>  See my reply below; not a good question.

2. When blit()ing into Eink[PIO] from a  1-bit-depth framebuffer (i.e. constructed with the framebuf.MONO_HLSB arg), white pixels are treated as transparent even when `key` is left undefined (-1). So if doing a partial display write over an existing image, only the black pixels are written: any existing black pixels will remain on the display and won't be clobbered by the white ones.  To get around this, I'm "scrubbing" the area first like so:

```        
        epd.rect(left, top , width, height, 1, f=True)
        epd.show()
        epd.rect(left, top, width, height, 0, f=True)
        epd.show()
```
before blitting the new image. Nothing less than that does the job. (For example, writing a white rectangle over existing black pixels doesn't seem to accomplish it, I have to turn the whole rectangle black first.) But this feels like a hack and is visible to the user. Is there way to blit it such that it writes both white and black pixels and no source pixel is treated as transparent? 


(*I've refactored your image-to-py conversion utility into a callable function, and using it on the web server that provides data to this device, generating these images dynamically from sensor data on request.)

BTW here's what it looks like. (Running instance connected only to a battery.) Yes, it's very warm here today:

![IMG_3159](https://github.com/phoreglad/pico-epaper/assets/6296075/7e6d6077-093a-4e40-95c6-845902788748)

**Reply by Cheddoleum** [reply]
(I'm sorry, on further consideration please ignore that first question: of course FrameBuffer is mostly a wrapper for memory that's already allocated; in this case, read from a web request and 'eval'd to turn it into a bytearray.  Reusing a FrameBuffer would not help here.)

**Reply by phoreglad** [reply]
Thanks for helping me make the library a bit better.

As for your question, I'm afraid it's just how the partial refresh works and there's not much I can do about it. Currently, I don't see any workaround using the available methods, but I have an idea of how it can be achieved. It would still be pretty hacky and be basically what you're doing now, but instead of using standard show() I'll have to add a method to update the screen's RAM without refreshing the display, so at least it wouldn't show for the user. It may not be the most elegant solution, but I'm not sure there's any other way.

**Reply by Cheddoleum** [reply]
Thanks again.  One thing I've done on the data-provider side is, after invoking your ImageToPy as a function from Flask, returning the images as application/octet-stream which the Pico converts directly to bytearrays -- still unavoidably making a copy, as FrameBuffer needs something mutable -- but it's 1/4th as big as a string representation.

**Reply by phoreglad** [reply]
Ok, I tried a few things but couldn't find a way to clear pre-existing image in partial mode. It seems partial mode requires a full screen refresh to apply changes and just updating RAM doesn't help. The way you currently clear the image is probably the only reliable approach.

**Reply by Cheddoleum** [reply]
Thanks for exploring that.

To anyone finding this thread, I wasn't thorough enough in testing my above claim about how to "scrub" a rectangle in partial mode before blit'ing a new image, and I'll correct the record:

It is **not** in fact necessary to draw a filled white rectangle before or after a black one in order to clear the existing pixels before a partial screen write. Drawing a filled black rectangle alone seems to suffice in practice to reset the pixels such that a subsequent B/W image can be blit'ed into that area without leaving residue from the black pixels that were originally there.   It's not obvious why it works this way, but it does in practice, so it's twice as quick. (And if you're drawing a white-on-black image, the process -- which takes place visibly in real time -- looks much better.)

**Comment by Cheddoleum** [comment]
Thanks again.  

Another question -- of the "just picking your brain" sort -- when/if you have a moment.  Right now I've got it set to do a full screen refresh (non-partial write) every 24 partial writes. This is a fairly high multiple of what Waveshare suggests.  But to be honest I'm really not clear on what problem one is trying to prevent by periodically doing a full rewrite every X writes. There is some suggestion of stuck pixels or something, but no real explanation of what the mechanism is or when and why it's a concern.  At minimum I really only need to clobber the whole screen when the day changes (see the photo above for the application). I'm strongly tempted to eliminate the other periodic full screen rewrites entirely, as I really don't know what problem it's solving.

**Reply by phoreglad** [reply]
To be honest, the e-ink displays are a bit of a magic and it's sometimes hard to find information on them. What is called a "partial update" in this driver (and the original one, where I took the name from) should really be called a "differential update" (sometimes also called "fast update") as only the pixels that changed from the previous image are updated. From what I understand doing so multiple times without full refresh can cause the particles in the screen to stick to one another instead of moving towards the electrodes or simply the wrong voltage being applied, which may lead to visible ghosts of the previous image. The exact mechanism is dependent on the screen type, I found some helpful explanations [here](https://electronics.stackexchange.com/questions/20276/why-does-flashing-prevent-ghosting-on-e-ink-displays). I'm not totally sure if it can lead to permanent damage to the display. Still, most manufacturers recommend that the full refresh is done every few partial/differential/fast updates to avoid potential problems and that's what I would stick to.

**Reply by Cheddoleum** [reply]
Thanks for that.  Interestingly, when you write a black rectangle to scrub an area, that rectangle actually briefly flashes all white first.  Which suggests to me that even in partial update mode, this particular device with its particular firmware used in this particular way, does a certain amount of custodial work to clear the pixels, at least in some use cases. In any event, I'm updating one region every minute for a time display, and I've extended the "full refresh" interval to something on the order of every 3 hours. It's been running 24/7 for a few days now. So far, not a hint of ghosting.
